content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+With this equivalency, checking if a given object is in a set becomes the
+*decision* problem of SAT - `x∈S`, exactly if `f[S](x)` evaluates to `true`.
+And checking if a set is empty becomes the *solution* problem of SAT - `S=∅`,
+exactly if `f[S]` is not satisfiable.
+
+With this, let us look at the specific sets we are interested in.
+
+## Basic interfaces as type sets
+
+Interfaces in Go have always described sets of types. For example, the interface
+
+```go
+type S interface {
+    X()
+    Y()
+    Z()
+}
+```
+
+is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.
+
+We also had a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):
+
+
+```go
+type S interface { X() }
+type T interface { Y() }
+type U interface {
+    S
+    T
+}
+```
+
+This expresses `T=S∩T` as an interface. We can also view the predicate “has a
+method `X()`” as a boolean variable and think of this as the formula `X∧Y`.
+
+Surprisingly, there is also a simple form of negation. It happens implicitly,
+because a type can not have two methods with the same name, but different
+types. So, implicitly, if a type has a method `X()`, it does *not* have a
+method `X() int`, for example[^5]:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+```
+
+This meant that you could express interfaces which could never be satisfied by
+any type (i.e. which described an empty type set):
+
+```go
+interface{ X; NotX }
+```
+
+[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).
+
+So, we pretty much had a language to build boolean formulas, where every
+formula had the form
+
+    X∧Y∧¬Z
+
+Checking if a type implements an interface meant to check if it is in the set
+described by that interfaces. Which meant checking if the corresponding formula
+is satisfied.
+
+And checking if an interface is invalid meant checking if the set it describes
+is empty. Which meant checking if the formula is *satisfiable*. This was easy,
+though. If you look closely, those formulas are in DNF and as we noted above,
+solving SAT on a formula given in DNF is easy.
+
+## Adding unions
+
+Go 1.18 extends the interface syntax. For our purposes, the important addition
+is the `|` operator:
+
+```go
+type S interface{
+    A | B
+}
+```
+
+This represents the set of all types which are in the *union* of the type sets
+`A` and `B` - that is, the set of all types which are in `A` *or* in `B` (or
+both).
+
+This means our language of expressible formulas now also includes a
+`∨`-operator. What's more, the form of our formula is now a *Conjunctive*
+normal form - every line is a term of `∨` and the lines are connected by `∧`:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+type Y interface { Y() }
+type NotY interface{ Y() int }
+type Z interface { Z() }
+type NotZ interface{ Z() int }
+
+// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+type S interface {
```

Comment:
I think you need to say somewhere that this is actually prohibited by the [current spec](https://go.dev/ref/spec#Interface_types):

> Implementation restriction: A union (with more than one term) cannot contain the [predeclared identifier](https://go.dev/ref/spec#Predeclared_identifiers) comparable or interfaces that specify methods, or embed comparable or interfaces that specify methods
 
This is kinda important, as it undermines the whole premise of this article - it means that calculating a type set is Go _isn't_ actually harder than one might think, with the current spec at least.

Reply:
I understand the argument, but to me, this is an important structural decision for the flow of the article. I'm trying to follow the reasoning behind the evolution of the design and I'm doing that piece by piece, not assuming things I haven't introduced. Talking about this statement and why it (hopefully) solves the problems requires already having talked about the union operator and embedding and it also requires knowing the problem.

What's more, I don't fully agree with the statement that calculating a type set is not harder than one might think. I believe it is still harder than one might think, as there are still subtleties involved and everything I mention still needs to be kept in mind. And while it *might* not be NP-complete anymore, we don't actually know that. At least as far as I know.

FWIW, I originally mentioned this in the introduction, but I couldn't get it to a point where I was happy with it. Once I removed it and decided to move it to its own section in the back, the overall structure I wanted became far more clear.

I see the criticism and I'm trying to acknowledging it by calling what I'm doing a bait-and-switch. But I'm claiming creative freedom here :)